make scheduler_perf stable

sanposhiho opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

- Some test cases are using non-exist namespaces. Add `CreateNamespaceOp` to create these namespaces before using them by Pods.
- Use `numPodsScheduledPerNamespace` to record all created namespaces 
- mark test as failed when cleanup failed
- retry when API returns generateName conflict. Related #107790

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

sanposhiho commented:
/sig scheduling
/cc @alculquicondor

sanposhiho commented:
/retest

denkensk commented on test/integration/scheduler_perf/config/performance-config.yaml:
```diff
@@ -23,19 +23,22 @@
 
 - name: SchedulingPodAntiAffinity
   workloadTemplate:
+  - opcode: createNamespaces
```
I'm a little curious that if we `createNamespaces` before `createPods` in the `workloadTemlate` strictly like you add here. Do the issues raised in https://github.com/kubernetes/kubernetes/pull/107700 still exist?

sanposhiho replied:
My first thought was:
If users don't specify `namespace` in `createPods`, scheduler_perf defines a namespace for Pod automatically. ​#107700 changes scheduler_perf to create namespaces that are automatically defined. 
On the other hand, the case users specify `namespace` in `createPods`, it means "scheduler_perf doesn't define a namespace for Pod automatically.". In that cases, I thought users should create namespace by themselves with `createNamespaces`.

---

But, on second thought, it would be more convenient if scheduler_perf creates namespaces even when users specify `namespace` in `createPods`. So, I'll revert these changes and change scheduler_perf to do that.

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -652,9 +652,9 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
+					err := client.CoreV1().Namespaces().Delete(context.TODO(), namespace, metav1.DeleteOptions{})
 					if err != nil {
-						b.Errorf("failed to delete namespace %v", namespace)
+						klog.Errorf("failed to delete namespace %v, err %v", namespace, err)
```
Why is this not a test failure (`b.Errorf`) anymore and merely something needs to be logged (`klog.Errorf`)?

If this is not supposed to occur, then I prefer `b.Errorf`.

sanposhiho replied:
I just follow other Cleanup functions in scheduler_perf, so no strong reason here.
I'll fix all cleanup functions as your advice.

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -652,9 +652,9 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
 				}
 				b.Cleanup(func() {
-					err := client.CoreV1().Namespaces().Delete(ctx, namespace, metav1.DeleteOptions{})
+					err := client.CoreV1().Namespaces().Delete(context.TODO(), namespace, metav1.DeleteOptions{})
```
The creation code tolerates already existing namespaces. Can that really happen?

If the code is called twice for the same namespace name, then it will only be created once, but two cleanup functions will be registered. Only the first one will succeed, the second or later ones will fail with a NotFound. That error needs to be ignored.

I don't know how many namespaces need to be created. If it's just a few, then the current approach with trying the operation and ignoring acceptable errors (in both places!) is okay. If it's many, then a `var namespaces map[string]bool` outside of the for loop could track which namespaces were already created. Only one cleanup function is needed then, registered outside of the loop, which then iterates over the created namespaces.

sanposhiho replied:
> The creation code tolerates already existing namespaces. Can that really happen?

Users can create namespace with `createNamespace` operation. So, it may happen.

> If the code is called twice for the same namespace name, then it will only be created once, but two cleanup functions will be registered. Only the first one will succeed, the second or later ones will fail with a NotFound. That error needs to be ignored.

It's a fair point. I think it's enough to change it to register only when no error from the creation of namespace.

pohly replied:
Right, that's simpler.

sanposhiho commented:
Let me mark this PR as WIP until I finish running scheduler_perf in my machine to confirm these changes are really valid. It is expected to take some time because I'm running all the test cases (and my machine isn't good enough.😓 )

/retitle [WIP] scheduler_perf: use context.TODO not to introduce context canceled

@pohly If you want to use this as soon as possible for your KEP, please use my branch (or add the same changes to your branch). I haven't been able to run all test cases through, but at least it works better than the latest master.

pohly commented:
> If you want to use this as soon as possible for your KEP, please use my branch (or add the same changes to your branch). I haven't been able to run all test cases through, but at least it works better than the latest master.

I had already started doing that and so far the changes in this PR look good.

sanposhiho commented:
/retitle [WIP] make scheduler_perf stable

alculquicondor commented:
is this still WIP?

sanposhiho commented:
My PC still running the test, but looks good so far. So, remove wip.

```
BenchmarkPerfScheduling/SchedulingBasic/500Nodes-8         	       1	25308198728 ns/op
E0126 22:41:31.253284   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingBasic/5000Nodes-8        	       1	166772786327 ns/op
E0126 22:47:00.715241   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingPodAntiAffinity/500Nodes-8         	       1	38579923020 ns/op
E0126 22:47:56.139666   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingPodAntiAffinity/5000Nodes-8        	       1	330302720578 ns/op
E0126 22:56:16.099025   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingSecrets/500Nodes-8                 	       1	26120268168 ns/op
E0126 22:56:58.844082   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingSecrets/5000Nodes-8                	       1	176444376599 ns/op
E0126 23:02:36.558913   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingInTreePVs/500Nodes-8               	       1	36659940590 ns/op
E0126 23:03:30.150591   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingInTreePVs/5000Nodes-8              	       1	260086004538 ns/op
E0126 23:10:33.873556   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingMigratedInTreePVs/500Nodes-8       	       1	57431408355 ns/op
E0126 23:11:48.139208   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingMigratedInTreePVs/5000Nodes-8      	       1	424683408185 ns/op
E0126 23:21:35.670723   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingCSIPVs/500Nodes-8                  	       1	55071270942 ns/op
E0126 23:22:47.996636   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingCSIPVs/5000Nodes-8                 	       1	410341150880 ns/op
E0126 23:32:25.666943   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingPodAffinity/500Nodes-8             	       1	40469554747 ns/op
E0126 23:33:23.028028   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingPodAffinity/5000Nodes-8            	       1	366041365844 ns/op
E0126 23:42:22.575085   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingPreferredPodAffinity/500Nodes-8    	       1	41353193508 ns/op
E0126 23:43:21.455594   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingPreferredPodAffinity/5000Nodes-8   	       1	350990577557 ns/op
E0126 23:52:07.394969   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
BenchmarkPerfScheduling/SchedulingPreferredPodAntiAffinity/500Nodes-8         	       1	40828008657 ns/op
E0126 23:53:06.415570   11730 controller.go:167] Found stale data, removed previous endpoints on kubernetes service, apiserver didn't exit successfully previously
```

/retitle make scheduler_perf stable

sanposhiho commented:
I think all errors in scheduler_perf except `Found stale data` error will be resolved by this PR. (And `Found stale date` will resolved by #107677)

sanposhiho commented:
I've confirmed that there are no new errors regarding this PR change. But, the two workloads `Preemption` and `PreemptionPV` are failed on my machine. Other workloads look good.

Anyway, I think this PR is worth merging. We can resolve errors in `Preemption` and `PreemptionPV` in other PRs.

pohly commented:
> I've confirmed that there are no new errors regarding this PR change. But, the two workloads Preemption and PreemptionPV are failed on my machine.

Those worked fine for me with your PR.

Can you squash your changes? I think they are ready for review and a LGTM.

pohly commented on test/utils/create_resources.go:
```diff
@@ -59,7 +61,14 @@ func CreatePodWithRetries(c clientset.Interface, namespace string, obj *v1.Pod)
+			return true, nil
+		}
+		if apierrors.IsAlreadyExists(err) {
+			if strings.Contains(err.Error(), errorGenerateNameConflict) {
```
One nit: checking for the error text is fragile. Can you instead check whether `obj.Name` is empty?

sanposhiho replied:
Sure. Fixed.

sanposhiho commented:
/retest

sanposhiho commented:
/retest

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -637,24 +639,30 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
+			} else {
 				b.Cleanup(func() {
-					err := client.CoreV1().Namespaces().Delete(ctx, namespace, metav1.DeleteOptions{})
+					err := client.CoreV1().Namespaces().Delete(context.TODO(), namespace, metav1.DeleteOptions{})
```
nit: `ctx.TODO()` implies that some future change is needed, in which case a comment explaining what that change is would be good.

But here I think that nothing needs to be changed anymore, so `ctx.Background()` seems more suitable.

sanposhiho replied:
Fixed.

pohly commented:
I've tested 0d297ab and it worked for me. I only have one nit left: https://github.com/kubernetes/kubernetes/pull/107771/files#r793492551

sanposhiho commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -637,24 +639,30 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
-				if err != nil && !apierrors.IsAlreadyExists(err) {
+			}
+			// create namespace in case it's not created yet.
+			_, err := client.CoreV1().Namespaces().Create(ctx, &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}, metav1.CreateOptions{})
```
I don't think this will slow down the test that much because the time to create a namespace is trivial compared to creating 5000 pods, etc. This change creates one namespace per one `createPods`, not creates 5000 namespaces for 5000 pods created.
So, for example, this change only creates two namespaces in `SchedulingBasic/5000Nodes` test.
https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/config/performance-config.yaml#L18

What do you think? (Sorry if you've already figured these things out and are giving advice. 🙏  )

denkensk replied:
https://github.com/kubernetes/kubernetes/pull/107771#discussion_r792342599

As I said upper. I also think it better to `createNamespaces` before `createPods` in the `workloadTemlate` other than create namespace in `createPods`.  This will help us to separate the process into different single op. Just personal advice. Thanks

sanposhiho replied:
Yeah, that's another design choice we can choose; let `createPods` create only Pod.
But, currently, scheduler_perf is designed in such a way that `createPods` makes related resources. We also create resources other than namespace in `createPods` like PV, PVC. 
So, personally, it doesn't seem strange to me that `createPods` create namespaces. But, no strong opinion here.

What do you think?

pohly replied:
> I don't think this will slow down the test that much because the time to create one namespace is trivial compared to creating 500~10000 pods. 

Even if it only *creates* one namespace for all pods, it *tries* for every single one of them, which causes additional load. One can argue that it is the job of a benchmark to generate load and see how the system handles it, but this approach here is unusual and thus not representative for real deployments.

How about tracking in the function which namespaces are known to exist? The prior to the `Create` call check that and skip the call when it isn't necessary anymore.

sanposhiho replied:
> How about tracking in the function which namespaces are known to exist? The prior to the Create call check that and skip the call when it isn't necessary anymore.

I am not against this. But,  if `skipWaitToCompletion` isn't `true`, namespaces should be unique to check if all newly created Pods are scheduled [here](https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler_perf/scheduler_perf_test.go#L897) by fetching all Pods in namespace, so the namespace creation will not be(should not be) skipped in most cases.

---

IIUC, the only reason we're creating namespace in each `createPods` is to check if all newly created Pods are scheduled. So, how about this idea:
- create one namespace when start scheduler_perf and use that namespace if `createPodsOp.Namespace == nil` instead of creating new namespace for each `createPods`.
  - if `createPodsOp.Namespace != nil`, the namespace won't be created.
- Pods created in the same `createPods` will have the same label.
- check if all newly created Pods are scheduled by fetching with label selector, instead of fetching all Pods in a namespace.

This idea seems to avoid all of the concerns raised in this thread.

alculquicondor replied:
> Even if it only creates one namespace for all pods, it tries for every single one of them, which causes additional load.

Exactly. It's overwhelming the API server, using QPS, etc.

I think your proposal is fine. I think some of the tests require multiple namespaces (like the ones for pod affinity). But they should be creating and setting the desired namespace instead of getting it implicitly.

sanposhiho commented:
@alculquicondor @pohly 
And I'd like to propose to remove `Preemption` and `PreemptionPV` tests for now. What do you think about this?

> Last night, I worked on investigating this configuration value QPS and Burst.
https://github.com/kubernetes/kubernetes/blob/04f747d09fd2f8fb486c90a85122c0cb7b2df782/test/integration/scheduler_perf/util.go#L86-L87
And found if make this value smaller (QPS = 500, Burst = 1000), the error `connect: can't assign requested address` doesn’t seem to occur… but, it is going to take a very long time to finish the test case Preemption/5000Nodes . This is because it seems to take 1 minute to schedule 1 Pod in that test case. And also, I saw many errors like following from the scheduler.
```
"Evaluation failed on nominated node" err="nodeinfo not found for node name \"node-0-2fdz8\"" pod="namespace-2/pod-575jj" node="node-0-2fdz8"
```
> I'm not sure if the scheduler is not capable of successfully completing this test, or if some improvements in  scheduler_perf implementation can solve these problems. Anyway, I think it is not a good idea to leave this unsuccessful test case, so I propose to remove it. (We can add it later when we find a fix that makes this test case successful.)
https://kubernetes.slack.com/archives/C09TP78DV/p1643340751974299?thread_ts=1642623117.030500&cid=C09TP78DV

alculquicondor commented:
Preemption is the most expensive operation in kube-scheduler. The test is designed to do a lot of preemptions. How big is the computer you are running the test on?

Perhaps you can try reducing the parameters locally if it's giving you issues, just to make sure it works, but we really want to keep this test.

pohly commented:
> Preemption is the most expensive operation in kube-scheduler. The test is designed to do a lot of preemptions. 
> How big is the computer you are running the test on?

I see the same PreemptionBasic and PreemptionPVs recorded as FAILED, and I am running on a pretty beefy desktop system (64GB RAM, 36 cores, albeit with hyperthreading). I have not checked why they fail, but lack of resources shouldn't be the problem.

pohly commented:
BTW, would it make sense to define smaller versions of SchedulingPreferredAffinityWithNSSelector, SchedulingRequiredPodAntiAffinityWithNSSelector, SchedulingPreferredAntiAffinityWithNSSelector, and SchedulingRequiredPodAffinityWithNSSelector?

Those are only defined for 5000 nodes and each takes over two minutes.

sanposhiho commented:
> How big is the computer you are running the test on?

32GB memory / 42.3GHz quad‑core 10th‑generation Intel Core i7

sanposhiho commented:
Apply this proposal.
https://github.com/kubernetes/kubernetes/pull/107771#discussion_r794316649

alculquicondor commented on test/integration/scheduler_perf/config/performance-config.yaml:
```diff
@@ -337,26 +349,29 @@
       labelValues: ["zone1"]
+  - opcode: createNamespaces
+    prefix: sched
+    count: 2
```
isn't just one enough here?

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_legacy_test.go:
```diff
@@ -22,6 +22,8 @@ import (
 	"time"
 
+	"k8s.io/apimachinery/pkg/labels"
+
```
no empty line

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -22,11 +22,16 @@ import (
 	"time"
 
+	"k8s.io/apimachinery/pkg/selection"
+
```
no empty lines

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -485,15 +494,15 @@ func (co churnOp) patchParams(w *workload) (realOp, error) {
-	// Namespaces to block on. Empty array or not specifying this field signifies
-	// that the barrier should block on all namespaces.
-	Namespaces []string
+	// IDs of createPodsOp to block on. Empty array or not specifying this field signifies
```
Do we really need this change?

It should be fine to just block based on namespaces

sanposhiho replied:
you are right, I reverted this change.

sanposhiho commented:
@alculquicondor 
Thanks for reviewing. Addressed all your suggestions. (I confirmed all test cases (except Preemption and PreemptionPV) work fine. )

sanposhiho commented:
/retest

sanposhiho commented:
@alculquicondor @denkensk 
Could you please take another look?

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -637,39 +643,31 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
-				})
 			}
+			// default createPodsOpID is opIndex.
+			id := strconv.Itoa(opIndex)
```
is this a leftover?

sanposhiho replied:
Sorry. I've changed it to appropriate comments.

denkensk commented on test/integration/scheduler_perf/scheduler_test.go:
```diff
@@ -25,6 +25,8 @@ import (
 	"time"
 
+	"k8s.io/apimachinery/pkg/labels"
+
```
Delete this blank line

denkensk commented on test/integration/scheduler_perf/util.go:
```diff
@@ -28,11 +28,14 @@ import (
 	"time"
 
+	"k8s.io/apimachinery/pkg/util/sets"
+
```
Delete this blank line

denkensk commented on test/utils/create_resources.go:
```diff
@@ -59,7 +59,14 @@ func CreatePodWithRetries(c clientset.Interface, namespace string, obj *v1.Pod)
+		if err == nil {
+			return true, nil
+		}
+		if apierrors.IsAlreadyExists(err) {
```
Can we define functions to avoid a lot of repetitive code?

denkensk commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -1115,7 +1135,7 @@ func (p *namespacePreparer) prepare() error {
 		n.Name = fmt.Sprintf("%s-%d", p.prefix, i)
 		if err := testutils.RetryWithExponentialBackOff(func() (bool, error) {
-			_, err := p.client.CoreV1().Namespaces().Create(context.TODO(), n, metav1.CreateOptions{})
+			_, err := p.client.CoreV1().Namespaces().Create(context.Background(), n, metav1.CreateOptions{})
```
Why change `context.TODO()` to `context.Background()`?

sanposhiho replied:
Because of this https://github.com/kubernetes/kubernetes/pull/107771#discussion_r793492551

denkensk commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -879,43 +877,59 @@ func getNodePreparer(prefix string, cno *createNodesOp, clientset clientset.Inte
 // lowest observed QPS of ~10 pods/sec, a 5000-node test should complete.
-func waitUntilPodsScheduledInNamespace(ctx context.Context, podInformer coreinformers.PodInformer, name string, namespace string, wantCount int) error {
+// If a given namespace is empty, it will check Pods with all label.
+func waitUntilPodsScheduled(ctx context.Context, podInformer coreinformers.PodInformer, name string, namespace string, label string, wantCount int) error {
```
From my personal opinion, the previous name is better. WDYT? Not mandatory
waitUntilPodsScheduled is for all namespaces in the list.
waitUntilPodsScheduledInNamespace for one namesapce

sanposhiho replied:
I don't have a strong opinion here, so revert the change of renaming 👍

denkensk reviewed (commented):
Thanks @sanposhiho  for your efforts.
Some comments.

sanposhiho commented:
@alculquicondor @denkensk 
Thanks for reviewing, I addressed all of your reviews. Please take a look again 🙇

sanposhiho commented:
/retest

sanposhiho commented:
/retest

sanposhiho commented:
/retest

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -637,39 +643,33 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
+			// indexStr is used for label.
+			// All pods which are created in this op will have this index in label,
+			// and we will use that label in labelselector to wait all Pods to be scheduled.
+			indexStr := strconv.Itoa(opIndex)
```
Didn't we decide to just wait for ALL pods in the namespace to be scheduled?

I'm just looking for simplicity here.

sanposhiho committed:
Make scheduler_perf stable

sanposhiho commented:
I raised the issue for `Preemption` and `PreemptionPV` workload fail.
https://github.com/kubernetes/kubernetes/issues/108308

alculquicondor commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -600,8 +600,17 @@ func runWorkload(b *testing.B, tc *testCase, w *workload) []DataItem {
 	nextNodeIndex := 0
+	// numPodsScheduledPerNamespace has all namespaces created in workload and the number of pods they (will) have.
+	// All namespaces listed in numPodsScheduledPerNamespace will be cleaned up.
+	numPodsScheduledPerNamespace := make(map[string]int)
```
if the int is not necessary, you can use `sets.String`

sanposhiho replied:
The number of Pods is used in [`barrierOp`](https://github.com/kubernetes/kubernetes/blob/7588a82bdb1f761ef2a591d18478b0145915af5d/test/integration/scheduler_perf/scheduler_perf_test.go#L488), so `int` is needed.
https://github.com/kubernetes/kubernetes/blob/7588a82bdb1f761ef2a591d18478b0145915af5d/test/integration/scheduler_perf/scheduler_perf_test.go#L804

sanposhiho committed:
fix by gofmt

sanposhiho commented:
@alculquicondor Could you please take another look?

alculquicondor reviewed (commented):
/lgtm
/approve

sanposhiho commented:
/assign @mrhohn

MrHohn reviewed (approved):
/approve
